### PR TITLE
feat: cancel pending assignment functionality

### PR DESCRIPTION
**Pull request by katrinan029**
# Description
Create a cancel functionality for pending assignments that the admin of an enterprise learner credit no longer wants the learner to enroll in. The learner will receive a cancellation email notification.

# Solution
Add a celery task to the send out a braze email campaign to the learner when the cancellation is successful.

[JIRA Ticket](https://2u-internal.atlassian.net/browse/ENT-7882)

### CODE REVIEW

**Review by kiram15 — COMMENTED** [review]
Looks good so far! Let me know if you need any help with testing but no other notes

File: enterprise_access/apps/api/tasks.py [hunk 1]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
+def send_cancel_email_for_pending_assignment(cancelled_assignment):
+    """
+    Send email via braze for cancelling pending assignment
+
+    Args:
+        cancelled_assignment: (string) the subsidy request uuid
+    """
+    learner_content_assignment_action = LearnerContentAssignmentAction(
+        uuid=cancelled_assignment.uuid, assignment=cancelled_assignment, action_type=AssignmentActions.CANCELLED
+    )
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:143]: You don't want to set the `uuid` here (and you wouldn't want it to share the uuid of the related Assignment, anyway) - the model will create a new one by default for you.

File: enterprise_access/apps/api/tasks.py [hunk 2]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
 
+@shared_task(base=LoggedTaskWithRetry)
+def send_cancel_email_for_pending_assignment(cancelled_assignment):
+    """
+    Send email via braze for cancelling pending assignment
+
+    Args:
+        cancelled_assignment: (string) the subsidy request uuid
+    """
+    learner_content_assignment_action = LearnerContentAssignmentAction(
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:141]: OK, so remember from another comment that we should be passing an assignment uuid into this task - pretend it's called `cancelled_assignment_uuid`.  What you'll want to do is to first in this task is fetch the assignment record based on that uuid:
```python
cancelled_assignment = LearnerContentAssignment.objects.get(cancelled_assignment_uuid)
```

File: enterprise_access/apps/api/tasks.py [hunk 3]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
+        uuid=cancelled_assignment.uuid, assignment=cancelled_assignment, action_type=AssignmentActions.CANCELLED
+    )
+
+    if braze_trigger_properties is None:
+        braze_trigger_properties = {}
+
+    braze_client_instance = BrazeApiClient()
+    lms_client = LmsApiClient()
+    enterprise_customer_uuid = learner_content_assignment_action.assignment_configuration.enterprise_customer_uuid
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:150]: The `assignment_configuration` is a member of the `LearnerContentAssignment` model, not the action model. You could instead get to the customer uuid by doing this:
```python
enterprise_customer_uuid = cancelled_assignment.assignment_configuration.enterprise_customer_uuid
```

File: enterprise_access/apps/api/tasks.py [hunk 4]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
+    enterprise_customer_data = lms_client.get_enterprise_customer_data(enterprise_customer_uuid)
+    lms_user_id = learner_content_assignment_action.assignment.lms_user_id
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:152]: nit, could just be 
```python
lms_user_id = cancelled_assignment.lms_user_id
```

File: enterprise_access/apps/api/tasks.py [hunk 5]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
+
+    try:
+        recipient = braze_client_instance.create_recipient(
+            user_email=cancelled_assignment.learner_email,
+            lms_user_id=cancelled_assignment.lms_user_id,
+        )
+        braze_trigger_properties["first_name"] = lms_user_id
+        braze_trigger_properties["organization"] = enterprise_customer_data.name
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:160]: `enterprise_customer_data` should be a plain dictionary here, you'll have to reference `name` as a key:
```python
blah = enterprise_customer_data['name']
```

File: enterprise_access/apps/api/tasks.py [hunk 6]
```diff
@@ -127,6 +130,50 @@ def send_notification_email_for_request(
...
+        braze_client_instance.send_campaign_message(
+            settings.BRAZE_ASSIGNMENT_CANCELLED_NOTIFICATION_CAMPAIGN,
+            recipients=[recipient],
+            trigger_properties=braze_trigger_properties,
+        )
+        learner_content_assignment_action.completed_at = datetime.now()
+        learner_content_assignment_action.save()
+    except Exception as exc:
+        logger.error(f"Unable to send email for {lms_user_id} due to exception: {exc}")
+        learner_content_assignment_action.error_reason = exc
```
Comment (iloveagent57) [inline enterprise_access/apps/api/tasks.py:171]: you'll have to pick a reason from `AssignmentActionErrors.CHOICES`

File: enterprise_access/apps/content_assignments/models.py [hunk 7]
```diff
@@ -283,6 +283,31 @@ def add_successful_notified_action(self):
...
+        """
+        Returns the first successful "cancelled" LearnerContentAssignmentActions for this assignment,
+        or None if no such record exists.
+        """
+        return self.actions.filter(
+            action_type=AssignmentActions.CANCELLED,
+            error_reason=None,
+        ).first()
+
+    def add_successful_cancelled_action(self):
```
Comment (iloveagent57) [inline enterprise_access/apps/content_assignments/models.py:296]: Hmm, did you mean to use this in your task to create a new cancelled action for the assignment?

File: enterprise_access/apps/content_assignments/api.py [hunk 8]
```diff
@@ -351,6 +354,9 @@ def cancel_assignments(assignments: Iterable[LearnerContentAssignment]) -> dict:
         assignment_to_cancel.state = LearnerContentAssignmentStateChoices.CANCELLED
 
     cancelled_assignments = _update_and_refresh_assignments(cancelable_assignments, ['state'])
+    for cancelled_assignment in cancelled_assignments:
+        send_cancel_email_for_pending_assignment(cancelled_assignment)
```
[the remark this answers is not in the document]
Comment (katrinan029) [reply enterprise_access/apps/content_assignments/api.py]: Thanks for this tip! Updated code to reflect this feedback.

File: enterprise_access/apps/content_assignments/constants.py [hunk 9]
```diff
@@ -33,11 +33,13 @@ class AssignmentActions:
     LEARNER_LINKED = 'learner_linked'
     NOTIFIED = 'notified'
     REMINDED = 'reminded'
+    CANCELLED = 'cancelled'
```
[the remark this answers is not in the document]
Comment (katrinan029) [reply enterprise_access/apps/content_assignments/constants.py]: Done!

File: enterprise_access/apps/api/tasks.py (hunk 1 above)
Comment (katrinan029) [reply enterprise_access/apps/api/tasks.py]: That makes sense. Removed the uuid here.

File: enterprise_access/apps/api/tasks.py (hunk 2 above)
Comment (katrinan029) [reply enterprise_access/apps/api/tasks.py]: Done!

File: enterprise_access/apps/content_assignments/models.py (hunk 7 above)
Comment (katrinan029) [reply enterprise_access/apps/content_assignments/models.py]: Removed this since I don't think we'll need it.

**Review by johnnagro — APPROVED** [review]

**Review by iloveagent57 — APPROVED** [review]
Great job on this!

**Pull request merged**